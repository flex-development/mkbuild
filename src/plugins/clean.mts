/**
 * @file Plugins - clean
 * @module mkbuild/plugins/clean
 */

import pathe from '@flex-development/pathe'
import { ok } from 'devlop'
import type {
  NormalizedOutputOptions,
  Plugin,
  PluginContext
} from 'rollup'

/**
 * Plugin name.
 *
 * @internal
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'mkbuild:clean'

/**
 * Cleaned directory cache.
 *
 * @internal
 *
 * @const {Set<string>} directories
 */
const directories: Set<string> = new Set<string>()

plugin.PLUGIN_NAME = PLUGIN_NAME
plugin.directories = directories

export default plugin

/**
 * Create an output directory cleaner plugin.
 *
 * @internal
 *
 * @see {@linkcode Plugin}
 *
 * @this {void}
 *
 * @return {Plugin}
 *  Output directory cleaner plugin
 */
function plugin(this: void): Plugin {
  return {
    name: PLUGIN_NAME,
    renderStart: { handler: renderStart, order: 'pre', sequential: true }
  }

  /**
   * Remove and recreate an output directory.
   *
   * @see https://rollupjs.org/plugin-development/#renderstart
   *
   * @async
   *
   * @this {PluginContext}
   *
   * @param {NormalizedOutputOptions} output
   *  Normalized output options
   * @return {Promise<undefined>}
   */
  async function renderStart(
    this: PluginContext,
    output: NormalizedOutputOptions
  ): Promise<undefined> {
    ok(typeof output.clean === 'boolean', 'expected `output.clean`')
    ok(typeof output.fs === 'object', 'expected `output.fs`')
    ok(typeof output.root === 'string', 'expected `output.root`')
    ok(typeof output.write === 'boolean', 'expected `output.write`')

    if (
      output.clean &&
      output.root &&
      output.write &&
      typeof output.dir === 'string'
    ) {
      directories.add(output.root)

      /**
       * Absolute path to output directory.
       *
       * @const {string} path
       */
      const path: string = pathe.join(output.root, output.dir)

      if (!directories.has(path)) {
        await output.fs.rm(path, { force: true, recursive: true })
        await output.fs.mkdir(path, { recursive: true })
      }

      directories.add(path)
    }

    return void this
  }
}
