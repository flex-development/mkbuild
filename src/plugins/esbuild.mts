/**
 * @file Plugins - esbuild
 * @module mkbuild/plugins/esbuild
 */

import formatEsbuildMessage from '#internal/format-esbuild-message'
import type WithUndefined from '#types/with-undefined'
import type {
  EsbuildOptions,
  Format
} from '@flex-development/mkbuild'
import pathe from '@flex-development/pathe'
import { ifelse, isNIL } from '@flex-development/tutils'
import { createFilter } from '@rollup/pluginutils'
import { ok } from 'devlop'
import type * as Esbuild from 'esbuild'
import type { Loader, LogLevel, TransformOptions, TsconfigRaw } from 'esbuild'
import type {
  MinimalPluginContext,
  ModuleInfo,
  NormalizedInputOptions,
  NormalizedOutputOptions,
  OutputBundle,
  Plugin,
  PluginContext,
  ProgramNode,
  RenderedChunk,
  RollupLog,
  LogLevel as RollupLogLevel,
  SourceDescription,
  TransformPluginContext,
  TransformResult
} from 'rollup'

/**
 * Plugin name.
 *
 * @internal
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'mkbuild:esbuild'

/**
 * esbuild module.
 *
 * @internal
 *
 * @var {typeof Esbuild | undefined} esbuild
 */
let esbuild: typeof Esbuild | undefined

plugin.PLUGIN_NAME = PLUGIN_NAME
export default plugin

/**
 * Plugin options.
 *
 * @internal
 *
 * @extends {EsbuildOptions}
 */
interface Options extends EsbuildOptions {
  /**
   * Module format.
   *
   * @see {@linkcode Format}
   */
  format: Format

  /**
   * Set the value of the `sourceRoot` field in a source map, which specifies
   * the path that all other paths in the source map are relative to. If this
   * field is not present, all paths in the source map are interpreted as being
   * relative to the directory containing the source map instead.
   *
   * @see https://esbuild.github.io/api/#source-root
   */
  sourceRoot?: URL | string | null | undefined

  /**
   * Include the `sourcesContent` field in sourcemaps.
   *
   * @see https://esbuild.github.io/api/#sources-content
   */
  sourcesContent?: boolean | null | undefined
}

/**
 * Create a plugin to transform files with esbuild.
 *
 * @see {@linkcode Options}
 * @see {@linkcode Plugin}
 *
 * @internal
 *
 * @this {void}
 *
 * @param {Options} [opts]
 *  esbuild options
 * @return {Plugin}
 *  esbuild transform plugin
 */
function plugin(this: void, opts: Options): Plugin {
  /**
   * Copied modules.
   *
   * @const {Set<ModuleInfo>} copied
   */
  const copied: Set<ModuleInfo> = new Set<ModuleInfo>()

  /**
   * Module id filter.
   *
   * @const {(id: unknown) => boolean} filter
   */
  const filter: (id: unknown) => boolean = createFilter(
    opts.include ?? /\.(?:(?:json)|(?:[cm]?[jt]s)|(?:[jt]sx?))$/,
    opts.exclude ?? undefined
  )

  /**
   * Raw tsconfig.
   *
   * @var {TsconfigRaw} tsconfigRaw
   */
  let tsconfigRaw: TsconfigRaw

  return {
    buildStart,
    generateBundle,
    name: PLUGIN_NAME,
    onLog,
    renderChunk,
    transform
  }

  /**
   * Get the raw tsconfig from `options`.
   *
   * @this {PluginContext}
   *
   * @param {NormalizedInputOptions} options
   *  Normalized input options
   * @return {undefined}
   */
  function buildStart(
    this: PluginContext,
    options: NormalizedInputOptions
  ): undefined {
    ok(typeof options.tsconfig === 'object', 'expected `options.tsconfig`')
    tsconfigRaw = options.tsconfig as TsconfigRaw
    return void this
  }

  /**
   * Add the content of copied modules to their chunks. Any sourcemaps generated
   * for copied modules will be removed.
   *
   * @see https://rollupjs.org/plugin-development/#generatebundle
   *
   * @this {PluginContext}
   *
   * @param {NormalizedOutputOptions} options
   *  Normalized output options
   * @param {OutputBundle} bundle
   *  Output bundle
   * @return {undefined}
   */
  function generateBundle(
    this: PluginContext,
    options: NormalizedOutputOptions,
    bundle: OutputBundle
  ): undefined {
    for (const output of Object.values(bundle)) {
      for (const info of copied) {
        if (output.type === 'chunk' && output.facadeModuleId === info.id) {
          ok(typeof info.code === 'string', 'expected `info.code`')

          output.code = info.code
          output.map = null

          if (output.sourcemapFileName !== null) {
            delete bundle[output.sourcemapFileName]
            output.sourcemapFileName = null
          }
        }
      }
    }

    return void bundle
  }

  /**
   * Skip `EMPTY_BUNDLE` logs.
   *
   * @see https://rollupjs.org/plugin-development/#onlog
   *
   * @this {MinimalPluginContext}
   *
   * @param {RollupLogLevel} level
   *  Rollup log level
   * @param {RollupLog} log
   *  Rollup log object
   * @return {false | undefined}
   *  `false` if log will be silenced, `undefined` otherwise
   */
  function onLog(
    this: MinimalPluginContext,
    level: RollupLogLevel,
    log: RollupLog
  ): false | undefined {
    return ifelse(log.code === 'EMPTY_BUNDLE', false, void log)
  }

  /**
   * Fix the name and file name of copied modules.
   *
   * @see https://rollupjs.org/plugin-development/#renderchunk
   *
   * @this {PluginContext}
   *
   * @param {string} code
   *  Chunk content
   * @param {RenderedChunk} chunk
   *  The chunk to process
   * @return {undefined}
   */
  function renderChunk(
    this: PluginContext,
    code: string,
    chunk: RenderedChunk
  ): undefined {
    for (const info of copied) {
      if (chunk.facadeModuleId === info.id) {
        chunk.fileName = pathe.changeExt(chunk.fileName, pathe.extname(info.id))
        chunk.name = pathe.changeExt(chunk.fileName, null)
        break
      }
    }

    return void chunk
  }

  /**
   * Transform `code` using the [esbuild transform api][esbuild-transform].
   *
   * [esbuild-transform]: https://esbuild.github.io/api/#transform
   *
   * @see https://rollupjs.org/plugin-development/#transform
   *
   * @todo `globalName`
   * @todo `ignoreAnnotations`
   * @todo `pure`
   * @todo `treeShaking`
   *
   * @this {TransformPluginContext}
   *
   * @param {string} code
   *  The code to transform
   * @param {string} id
   *  The id of the module containing `code`
   * @return {Promise<TransformResult>}
   *  Transform result
   */
  async function transform(
    this: TransformPluginContext,
    code: string,
    id: string
  ): Promise<TransformResult> {
    /**
     * Transform result.
     *
     * @var {WithUndefined<Partial<SourceDescription>> | undefined} result
     */
    let result: WithUndefined<Partial<SourceDescription>> | undefined

    if (filter(id)) {
      /**
       * AST of transformed code.
       *
       * @var {ProgramNode | undefined} ast
       */
      let ast: ProgramNode | undefined

      /**
       * Content loader.
       *
       * @var {Loader | undefined} loader
       */
      let loader: Loader | undefined = pathe.extToValue(pathe.basename(id), {
        '.cjs': 'js',
        '.cts': 'ts',
        '.d.cts': 'copy',
        '.d.mts': 'copy',
        '.d.ts': 'copy',
        '.js': 'js',
        '.json': 'copy',
        '.jsx': 'jsx',
        '.mjs': 'js',
        '.mts': 'ts',
        '.ts': 'ts',
        '.tsx': 'tsx',
        ...opts.loader
      })

      // note: rather than just be a no-op, or work as `empty`, esbuild throws
      // an annoying error when the `copy` loader is used with `transform`.
      if (loader === 'copy') {
        /**
         * Module info.
         *
         * @const {ModuleInfo | null} info
         */
        const info: ModuleInfo | null = this.getModuleInfo(id)

        if (info) {
          copied.add(info)
          info.code = code
        }

        ast = {
          body: [],
          end: 0,
          sourceType: 'module',
          start: 0,
          type: 'Program'
        }

        loader = 'empty'
      }

      const { transform } = esbuild ??= await import('esbuild')

      /**
       * Log level overrides.
       *
       * @const {Record<string, LogLevel>} logOverride
       */
      const logOverride: Record<string, LogLevel> = { ...opts.logOverride }

      ok(typeof opts.sourcesContent === 'boolean', 'expected `sourcesContent`')

      const {
        code: transformed,
        map,
        legalComments,
        mangleCache,
        warnings
      } = await transform(code, {
        charset: opts.charset ?? undefined,
        color: false,
        define: opts.define ?? undefined,
        drop: opts.drop ?? undefined,
        dropLabels: opts.dropLabels ?? undefined,
        format: opts.format,
        jsx: opts.jsx ?? undefined,
        jsxDev: opts.jsxDev ?? undefined,
        jsxFactory: opts.jsxFactory ?? undefined,
        jsxFragment: opts.jsxFragment ?? undefined,
        jsxImportSource: opts.jsxImportSource ?? undefined,
        jsxSideEffects: opts.jsxSideEffects ?? undefined,
        keepNames: opts.keepNames ?? undefined,
        legalComments: 'external',
        lineLimit: opts.lineLimit ?? undefined,
        loader,
        logLevel: 'silent',
        logLimit: 0,
        logOverride: Object
          .entries(logOverride)
          .reduce<Record<string, LogLevel>>((acc, [key, level]) => {
            acc[key] = ifelse(level === 'silent', level, 'warning')
            return acc
          }, {}),
        mangleCache: opts.mangleCache ?? undefined,
        mangleProps: opts.mangleProps ?? undefined,
        mangleQuoted: opts.mangleQuoted ?? undefined,
        minify: opts.minify ?? undefined,
        minifyIdentifiers: opts.minifyIdentifiers ?? undefined,
        minifySyntax: opts.minifySyntax ?? undefined,
        minifyWhitespace: opts.minifyWhitespace ?? undefined,
        platform: opts.platform ?? undefined,
        reserveProps: opts.reserveProps ?? undefined,
        sourceRoot: ifelse(
          isNIL(opts.sourceRoot),
          undefined,
          String(opts.sourceRoot)
        ),
        sourcefile: id,
        sourcemap: true,
        sourcesContent: opts.sourcesContent,
        supported: {
          'dynamic-import': true,
          'import-meta': true,
          ...opts.supported
        },
        target: typeof opts.target === 'object'
          ? [...new Set(opts.target)]
          : opts.target,
        tsconfigRaw: opts.tsconfigRaw ?? tsconfigRaw
      } as TransformOptions)

      for (const message of warnings) {
        /**
         * Message formatted as rollup log.
         *
         * @const {RollupLog} log
         */
        const log: RollupLog = formatEsbuildMessage(
          message,
          code,
          id,
          PLUGIN_NAME,
          logOverride[message.id]
        )

        // log message if not silenced
        log.level && this[log.level](log, log.loc)
      }

      result = {
        ast,
        code: typeof ast === 'object' ? code : transformed,
        map,
        meta: { legalComments, mangleCache, warnings }
      }
    }

    return result as TransformResult
  }
}
