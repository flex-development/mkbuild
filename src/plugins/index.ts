/**
 * @file Entry Point - Plugins
 * @module mkbuild/plugins
 */

export { default as clean } from './clean/plugin'
export { default as createRequire } from './create-require/plugin'
export { default as decorators } from './decorators/plugin'
export { default as dts } from './dts/plugin'
export { default as filter } from './filter/plugin'
export { default as fullySpecified } from './fully-specified/plugin'
export { default as pkgtype } from './pkgtype/plugin'
export { default as tsconfigPaths } from './tsconfig-paths/plugin'
export { default as write } from './write/plugin'
