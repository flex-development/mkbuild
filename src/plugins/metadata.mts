/**
 * @file Plugins - metadata
 * @module mkbuild/plugins/metadata
 */

import { ksort } from '@flex-development/tutils'
import type {
  NormalizedOutputOptions,
  OutputBundle,
  Plugin,
  PluginContext
} from 'rollup'

/**
 * Plugin name.
 *
 * @internal
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'mkbuild:metadata'

plugin.PLUGIN_NAME = PLUGIN_NAME
export default plugin

/**
 * Create a build metadata plugin.
 *
 * @see {@linkcode Plugin}
 *
 * @internal
 *
 * @this {void}
 *
 * @return {Plugin}
 *  Build metadata plugin
 */
function plugin(this: void): Plugin {
  return { generateBundle, name: PLUGIN_NAME }

  /**
   * Attach additional metadata to output assets and chunks.
   *
   * @see https://rollupjs.org/plugin-development/#generatebundle
   *
   * @this {PluginContext}
   *
   * @param {NormalizedOutputOptions} options
   *  Normalized output options
   * @param {OutputBundle} bundle
   *  Output bundle
   * @return {undefined}
   */
  function generateBundle(
    this: PluginContext,
    options: NormalizedOutputOptions,
    bundle: OutputBundle
  ): undefined {
    for (const output of Object.values(bundle)) {
      /**
       * Output file contents.
       *
       * @const {Uint8Array | string} contents
       */
      const contents: Uint8Array | string = 'code' in output
        ? output.code
        : output.source

      output.bytes = Buffer.byteLength(contents)
      ksort(output)
    }

    return void this
  }
}
