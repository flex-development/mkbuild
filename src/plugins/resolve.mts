/**
 * @file Plugins - resolve
 * @module mkbuild/plugins/resolve
 */

import VIRTUAL_MODULE_RE from '#internal/virtual-module-re'
import type WithUndefined from '#types/with-undefined'
import { isBuiltin } from '@flex-development/is-builtin'
import type {
  ResolveIdOptions,
  ResolveOptions
} from '@flex-development/mkbuild'
import {
  isRelativeSpecifier,
  resolveAlias,
  resolveModule,
  toRelativeSpecifier,
  type ResolveModuleOptions
} from '@flex-development/mlly'
import pathe from '@flex-development/pathe'
import { ok } from 'devlop'
import type {
  NormalizedInputOptions,
  PartialResolvedId,
  Plugin,
  PluginContext
} from 'rollup'

/**
 * Plugin name.
 *
 * @internal
 *
 * @const {string} PLUGIN_NAME
 */
const PLUGIN_NAME: string = 'mkbuild:resolve'

plugin.PLUGIN_NAME = PLUGIN_NAME
export default plugin

/**
 * Plugin API.
 *
 * @see {@linkcode ResolveModuleOptions}
 *
 * @internal
 *
 * @extends {ResolveModuleOptions}
 */
interface Api extends ResolveModuleOptions {
  /**
   * URL of root directory.
   *
   * @override
   */
  cwd: URL
}

/**
 * Create a module resolver plugin.
 *
 * @todo `makeAbsoluteExternalsRelative` support (?)
 *
 * @internal
 *
 * @see {@linkcode Api}
 * @see {@linkcode Plugin}
 * @see {@linkcode ResolveOptions}
 *
 * @this {void}
 *
 * @param {ResolveOptions | null | undefined} [options]
 *  Module resolve options
 * @return {Plugin<Api>}
 *  Module resolver plugin
 */
function plugin(
  this: void,
  options?: ResolveOptions | null | undefined
): Plugin<Api> {
  /**
   * Plugin api.
   *
   * @const {Api} api
   */
  const api: Api = { ...options } as Api

  return {
    api,
    buildStart,
    name: PLUGIN_NAME,
    resolveId
  }

  /**
   * Initialize the plugin api.
   *
   * @this {PluginContext}
   *
   * @param {NormalizedInputOptions} options
   *  Normalized input options
   * @return {undefined}
   */
  function buildStart(
    this: PluginContext,
    options: NormalizedInputOptions
  ): undefined {
    ok(typeof options.root === 'string', 'expected `root`')

    api.cwd = pathe.pathToFileURL(options.root)

    api.extensions ??= [
      '.mjs',
      '.js',
      '.mts',
      '.ts',
      '.jsx',
      '.tsx',
      '.json'
    ]

    api.fs = options.fs

    return void this
  }

  /**
   * Resolve `id`.
   *
   * > ðŸ‘‰ **Note**: Does nothing if `id` follows virtual module naming
   * > conventions.
   *
   * @see https://rollupjs.org/plugin-development/#resolveid
   *
   * @async
   *
   * @this {PluginContext}
   *
   * @param {string} id
   *  The module specifier to resolve
   * @param {string | undefined} importer
   *  The parent module path
   * @param {ResolveIdOptions} options
   *  Hook options
   * @return {Promise<PartialResolvedId | undefined>}
   *  Resolve result
   * @throws {unknown}
   */
  async function resolveId(
    this: PluginContext,
    id: string,
    importer: string | undefined,
    options: ResolveIdOptions
  ): Promise<PartialResolvedId | undefined> {
    /**
     * Resolve result.
     *
     * @var {WithUndefined<PartialResolvedId> | undefined} resolved
     */
    let resolved: WithUndefined<PartialResolvedId> | undefined

    if (!VIRTUAL_MODULE_RE.test(id)) {
      ok(api.cwd, 'expected `api.cwd`')

      /**
       * URL of parent module.
       *
       * @const {URL} parent
       */
      const parent: URL = importer ? pathe.pathToFileURL(importer) : api.cwd

      if (isBuiltin(id)) {
        resolved = {
          external: true,
          id: String(await resolveModule(id, parent))
        }
      } else {
        /**
         * Specifier of aliased module.
         *
         * @const {string | null} url
         */
        const aliased: string | null = resolveAlias(id, {
          ...api,
          absolute: false,
          parent
        })

        // reset `id` if path alias match was found
        id = aliased ?? id

        /**
         * URL of resolved module.
         *
         * @const {URL} url
         */
        const url: URL = await resolveModule(id, parent, api)

        resolved = {
          external: !options.isEntry,
          id: isRelativeSpecifier(id) ? toRelativeSpecifier(url, parent) : id
        }
      }
    }

    return resolved as PartialResolvedId
  }
}
