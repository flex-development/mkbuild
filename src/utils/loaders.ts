/**
 * @file Utilities - loaders
 * @module mkbuild/utils/loaders
 */

import type * as pathe from '@flex-development/pathe'
import type * as esbuild from 'esbuild'

/**
 * Returns an esbuild [build api][1] [`loader`][2] configuration.
 *
 * [1]: https://esbuild.github.io/api/#build-api
 * [2]: https://esbuild.github.io/api/#loader
 *
 * @param {esbuild.Format} [format='esm'] - Output file format
 * @param {boolean} [bundle=false] - Bundling enabled?
 * @return {Record<pathe.Ext, esbuild.Loader>} `loader` configuration
 */
const loaders = (
  format: esbuild.Format = 'esm',
  bundle: boolean = false
): Record<pathe.Ext, esbuild.Loader> => ({
  '.cjs': format === 'cjs' && !bundle ? 'copy' : 'js',
  '.css': bundle ? 'css' : 'copy',
  '.cts': 'ts',
  '.d.cts': 'copy',
  '.d.mts': 'copy',
  '.d.ts': 'copy',
  '.data': bundle ? 'binary' : 'copy',
  '.eot': 'copy',
  '.gif': 'copy',
  '.jpeg': 'copy',
  '.jpg': 'copy',
  '.js': 'js',
  '.json': bundle ? 'json' : 'copy',
  '.json5': bundle ? 'json' : 'copy',
  '.jsonc': bundle ? 'json' : 'copy',
  '.jsx': 'jsx',
  '.mjs': format === 'esm' && !bundle ? 'copy' : 'js',
  '.mts': 'ts',
  '.otf': 'copy',
  '.png': 'copy',
  '.svg': 'copy',
  '.ts': 'ts',
  '.tsx': 'tsx',
  '.txt': bundle ? 'text' : 'copy',
  '.woff': 'copy',
  '.woff2': 'copy'
})

export default loaders
