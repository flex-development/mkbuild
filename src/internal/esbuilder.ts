/**
 * @file Internal - esbuilder
 * @module mkbuild/internal/esbuilder
 */

import type { Entry, Result, SourceFile } from '#src/interfaces'
import * as mkp from '#src/plugins'
import type { FileSystemAdapter, OutputMetadata } from '#src/types'
import fsa from '#src/utils/fs'
import IGNORE from '#src/utils/ignore-patterns'
import loaders from '#src/utils/loaders'
import { EXT_DTS_REGEX } from '@flex-development/ext-regex'
import * as mlly from '@flex-development/mlly'
import * as pathe from '@flex-development/pathe'
import * as esbuild from 'esbuild'
import regexp from 'escape-string-regexp'
import fg from 'fast-glob'
import { omit } from 'radash'

/**
 * Builds `entry` using the [esbuild build API][1].
 *
 * [1]: https://esbuild.github.io/api/#build-api
 *
 * @todo ensure `entry.source` is a directory when bundling is disabled
 * @todo ensure `entry.source` is a file (with extname) when bundling is enabled
 *
 * @async
 *
 * @param {Partial<Entry>} entry - Build entry object
 * @param {Partial<FileSystemAdapter>} [fs=fsa] - Custom file system methods
 * @return {Promise<[esbuild.Metafile, Result[]]>} Metafile and build results
 * @throws {esbuild.BuildFailure}
 */
const esbuilder = async (
  entry: Partial<Entry>,
  fs: Partial<FileSystemAdapter> = fsa
): Promise<[esbuild.Metafile, Result[]]> => {
  const {
    bundle = false,
    conditions = bundle ? [] : mlly.CONDITIONS,
    createRequire: insertRequire,
    cwd = '.',
    dts: declaration,
    ext = '.mjs',
    external = [],
    format = 'esm',
    ignore = IGNORE,
    loader = {},
    name = '[name]',
    outExtension = {},
    outdir = 'dist',
    pattern = '**',
    platform = '',
    plugins = [],
    resolveExtensions = mlly.RESOLVE_EXTENSIONS,
    source = bundle ? 'src/index' : 'src',
    tsconfig = '',
    write = false,
    outbase = bundle ? pathe.parse(source).dir : source,
    ...options
  } = entry

  /**
   * Absolute path to current working directory.
   *
   * @const {string} absWorkingDir
   */
  const absWorkingDir: string =
    pathe.resolve(cwd).replace(/\/$/, '') + pathe.sep

  /**
   * Relative paths to source files.
   *
   * **Note**: Files are relative to {@linkcode source} when {@linkcode bundle}
   * is `false`. When enabled, files are relative to {@linkcode absWorkingDir}.
   *
   * @const {string[]} files
   */
  const files: string[] = await fg(bundle ? source : pattern, {
    braceExpansion: true,
    caseSensitiveMatch: true,
    cwd: pathe.join(absWorkingDir, bundle ? '' : source),
    dot: true,
    extglob: true,
    fs: omit(fs, ['readdir']),
    globstar: true,
    ignore: [...new Set(ignore)],
    objectMode: false,
    onlyFiles: true,
    stats: false,
    throwErrorOnBrokenSymbolicLink: false,
    unique: true
  })

  /**
   * Source file objects.
   *
   * @const {SourceFile[]} sources
   */
  const sources: SourceFile[] = files.map(file => {
    /**
     * Absolute path to source file.
     *
     * @const {string} path
     */
    const path: string = pathe.join(absWorkingDir, source, bundle ? '' : file)

    /**
     * File extension of source file.
     *
     * @var {string} ext
     */
    let ext: string = pathe.extname(file)

    // fix file extension if source file is typescript declaration file
    /* c8 ignore next */ EXT_DTS_REGEX.test(file) && (ext = '.d' + ext)

    return {
      ext: ext as pathe.Ext,
      file: bundle
        ? // adds outbase support for bundles (esbuild only uses outbase for
          // multiple entry points); @see https://esbuild.github.io/api/#outbase
          file.replace(new RegExp(`^${regexp(outbase)}`), '').replace(/^\//, '')
        : file,
      path
    }
  })

  // add fully-specified plugin
  if (format === 'esm' || ext === '.cjs') plugins.unshift(mkp.fullySpecified())

  // add tsconfig-paths plugin
  tsconfig && plugins.unshift(mkp.tsconfigPaths())

  // add dts plugin
  if (declaration as boolean) plugins.unshift(mkp.dts())

  // add create-require plugin
  if ((bundle && format === 'esm' && platform === 'node') || insertRequire) {
    plugins.unshift(mkp.createRequire())
  }

  // add decorators plugin
  tsconfig && plugins.unshift(mkp.decorators())

  // add output file writer plugin
  if (write) {
    plugins.push(
      mkp.write({
        ...fs,
        filter: declaration === 'only' ? EXT_DTS_REGEX : undefined
      })
    )
  }

  // remove unsupported options that are not overridden
  Reflect.deleteProperty(options, 'incremental')
  Reflect.deleteProperty(options, 'outfile')
  Reflect.deleteProperty(options, 'stdin')
  Reflect.deleteProperty(options, 'watch')

  // build source files
  const { errors, metafile, outputFiles, warnings } = await esbuild.build({
    ...options,
    absWorkingDir,
    bundle,
    conditions: [...new Set(conditions)],
    entryNames: `[dir]/${name}`,
    entryPoints: sources.reduce<Record<string, string>>((acc, src) => {
      const { ext, file } = src

      /**
       * Regex pattern matching {@linkcode ext}.
       *
       * @const {RegExp} ext_re
       */
      const ext_re: RegExp = new RegExp(`${regexp(ext)}$`)

      acc[file.replace(ext_re, '')] = pathe.join(source, bundle ? '' : file)
      return acc
    }, {}),
    external: bundle ? [...new Set(external)] : [],
    format,
    loader: { ...loaders(format, bundle), ...loader },
    metafile: true,
    outExtension: { ...outExtension, '.js': ext },
    outbase,
    outdir,
    platform: platform as esbuild.Platform,
    plugins,
    resolveExtensions: [...new Set(resolveExtensions)].map(ext => {
      return pathe.formatExt(ext)
    }),
    tsconfig,
    write: false
  })

  return [
    metafile,
    outputFiles
      .map(({ contents, path, text }) => {
        /**
         * Relative path to output file.
         *
         * **Note**: Relative to {@linkcode absWorkingDir}.
         *
         * @const {string} outfile
         */
        const outfile: string = path.replace(
          new RegExp(`^${regexp(absWorkingDir)}`),
          ''
        )

        /**
         * Output metadata.
         *
         * @const {OutputMetadata} metadata
         */
        const metadata: OutputMetadata = metafile.outputs[outfile]!

        return {
          bytes: metadata.bytes,
          contents,
          entryPoint: metadata.entryPoint,
          errors,
          exports: [...new Set(metadata.exports)],
          imports: metadata.imports,
          inputs: metafile.inputs,
          outfile,
          path,
          text,
          warnings
        }
      })
      .filter(res => {
        return declaration === 'only' ? EXT_DTS_REGEX.test(res.path) : true
      })
  ]
}

export default esbuilder
