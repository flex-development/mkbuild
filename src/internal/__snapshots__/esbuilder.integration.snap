// Vitest Snapshot v1

exports[`integration:internal/esbuilder > esbuild > cjs > should create bundle in cjs format 1`] = `
[
  {
    "entryPoint": "find-uniq.cts",
    "errors": [],
    "outfile": "dist/find-uniq.cjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/find-uniq/dist/find-uniq.cjs",
    "warnings": [],
  },
]
`;

exports[`integration:internal/esbuilder > esbuild > cjs > should do transpilation in cjs format 1`] = `
[
  {
    "entryPoint": "find-uniq.cts",
    "errors": [],
    "outfile": "dist/find-uniq.cjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/find-uniq/dist/find-uniq.cjs",
    "warnings": [],
  },
]
`;

exports[`integration:internal/esbuilder > esbuild > esm > should create bundle in esm format 1`] = `
[
  {
    "entryPoint": undefined,
    "errors": [],
    "outfile": "dist/index.mjs.map",
    "path": "\${process.cwd()}/__fixtures__/pkg/reverse/dist/index.mjs.map",
    "warnings": [],
  },
  {
    "entryPoint": "src/index.mts",
    "errors": [],
    "outfile": "dist/index.mjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/reverse/dist/index.mjs",
    "warnings": [],
  },
]
`;

exports[`integration:internal/esbuilder > esbuild > esm > should do transpilation in esm format 1`] = `
[
  {
    "entryPoint": "src/index.mts",
    "errors": [],
    "outfile": "dist/index.mjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/reverse/dist/index.mjs",
    "warnings": [],
  },
  {
    "entryPoint": "src/index.mts",
    "errors": [],
    "outfile": "dist/index.d.mts",
    "path": "\${process.cwd()}/__fixtures__/pkg/reverse/dist/index.d.mts",
    "warnings": [],
  },
  {
    "entryPoint": "src/max-value.ts",
    "errors": [],
    "outfile": "dist/max-value.mjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/reverse/dist/max-value.mjs",
    "warnings": [],
  },
  {
    "entryPoint": "src/max-value.ts",
    "errors": [],
    "outfile": "dist/max-value.d.mts",
    "path": "\${process.cwd()}/__fixtures__/pkg/reverse/dist/max-value.d.mts",
    "warnings": [],
  },
  {
    "entryPoint": "src/min-value.ts",
    "errors": [],
    "outfile": "dist/min-value.mjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/reverse/dist/min-value.mjs",
    "warnings": [],
  },
  {
    "entryPoint": "src/min-value.ts",
    "errors": [],
    "outfile": "dist/min-value.d.mts",
    "path": "\${process.cwd()}/__fixtures__/pkg/reverse/dist/min-value.d.mts",
    "warnings": [],
  },
  {
    "entryPoint": "src/reverse.mts",
    "errors": [],
    "outfile": "dist/reverse.mjs",
    "path": "\${process.cwd()}/__fixtures__/pkg/reverse/dist/reverse.mjs",
    "warnings": [],
  },
  {
    "entryPoint": "src/reverse.mts",
    "errors": [],
    "outfile": "dist/reverse.d.mts",
    "path": "\${process.cwd()}/__fixtures__/pkg/reverse/dist/reverse.d.mts",
    "warnings": [],
  },
]
`;

exports[`integration:internal/esbuilder > esbuild > iife > should create bundle in iife format 1`] = `
[
  {
    "entryPoint": "buddy.js",
    "errors": [],
    "outfile": "dist/buddy.js",
    "path": "\${process.cwd()}/__fixtures__/pkg/buddy/dist/buddy.js",
    "warnings": [],
  },
]
`;

exports[`integration:internal/esbuilder > esbuild > iife > should do transpilation in iife format 1`] = `
[
  {
    "entryPoint": "buddy.js",
    "errors": [],
    "outfile": "dist/buddy.js",
    "path": "\${process.cwd()}/__fixtures__/pkg/buddy/dist/buddy.js",
    "warnings": [],
  },
]
`;
